src/platform/nrfconnect/FactoryDataProvider.cpp

```diff
@@ -95,9 +89,95 @@ CHIP_ERROR FactoryDataProvider<FlashFactoryData>::Init()
+        error = CHIP_NO_ERROR;
+    }
+    else if (error != CHIP_NO_ERROR)
+    {
+        ChipLogError(DeviceLayer, "Failed to protect the factory data partition.");
+        return error;
+    }
+
+    return error;
 }
 
+#ifdef CONFIG_CHIP_CRYPTO_PSA
+template <class FlashFactoryData>
+CHIP_ERROR FactoryDataProvider<FlashFactoryData>::MoveDACPrivateKeyToSecureStorage(uint8_t * factoryDataPartition,
+                                                                                   size_t factoryDataSize)
+{
+
+    CHIP_ERROR err         = CHIP_NO_ERROR;
+    void * factoryDataBuff = nullptr;
+
+    if (!factoryDataPartition || factoryDataSize == 0)
+    {
+        return CHIP_ERROR_INVALID_ARGUMENT;
+    }
+
+    // Check if factory data contains DAC private key
+    if (mFactoryData.dacPrivateKeyPresent)
+    {
+        psa_key_attributes_t attributes = PSA_KEY_ATTRIBUTES_INIT;
+        // If there is the new DAC private key present in the factory data set and also there is
+        // the existing one in ITS NVM storage, then it is a security violation
+        // TODO: Consider other reactions to this violation: blocking the application, factory reset, etc.
+        if (psa_get_key_attributes(mDACPrivKeyId, &attributes) != PSA_SUCCESS)
+        {
+            ChipLogDetail(DeviceLayer, "Found DAC Private Key in factory data set. Copying to ITS...");
+
+            psa_reset_key_attributes(&attributes);
+            psa_set_key_type(&attributes, PSA_KEY_TYPE_ECC_KEY_PAIR(PSA_ECC_FAMILY_SECP_R1));
+            psa_set_key_bits(&attributes, kDACPrivateKeyLength * 8);
+            psa_set_key_algorithm(&attributes, PSA_ALG_ECDSA(PSA_ALG_SHA_256));
+            psa_set_key_lifetime(&attributes, PSA_KEY_LIFETIME_PERSISTENT);
+            psa_set_key_id(&attributes, mDACPrivKeyId);
+            psa_set_key_usage_flags(&attributes, PSA_KEY_USAGE_EXPORT | PSA_KEY_USAGE_SIGN_MESSAGE);
+
+            VerifyOrExit(psa_import_key(&attributes, reinterpret_cast<uint8_t *>(mFactoryData.dac_priv_key.data),
+                                        kDACPrivateKeyLength, &mDACPrivKeyId) == PSA_SUCCESS,
+                         err = CHIP_ERROR_INTERNAL);
+        }
+
+        memset(mFactoryData.dac_priv_key.data, 0xff, mFactoryData.dac_priv_key.len);
+        mFactoryData.dacPrivateKeyPresent = false;
+
+        // Allocate needed memory space to perform removal and moving DAC private key
+        size_t alignedSize = mFactoryData.actualSize + (sizeof(uint32_t) - (mFactoryData.actualSize % sizeof(uint32_t)));
+        factoryDataBuff    = chip::DeviceLayer::Malloc::Calloc(alignedSize, sizeof(uint8_t));
+
+        if (!factoryDataBuff)
+        {
+            ChipLogError(DeviceLayer, "Could not allocate memory for removing DAC private key!");
+            return CHIP_ERROR_NO_MEMORY;
+        }
+
+        VerifyOrExit(SerializeFactoryData(reinterpret_cast<uint8_t *>(factoryDataBuff), mFactoryData.actualSize, &mFactoryData),
```

Comment:
Why do we need to serialize the factory data? Wouldn't it be much easier to simply set the DAC private key to ffff... and save the original CBOR with just this small modification? Then we wouldn't need to pull ZCBOR encoding code and it should be less error-prone (we don't have to keep the parsing and serializing code in sync in the future).

Reply:
Right, I thought that we may need it in the future, but after consideration, I agree that we won't need to modify the factory data in any other way than removing DAC. So I will simply overwrite the dac priv key and flash the new factory data set.